### DISCUSSION TITLE
Could not locate Comsol 5.2 on Windows

### DISCUSSION THREAD

**Discussion by Darkforg** [MPh-py/MPh#151] [General]
Hello!

As seen in #150 I have the same problem, even though comsol is put on the default folder, mph cannot locate the installation. I understand (as I saw in documentaiton) that this is probably because of comsol's version (5.2). It is also classkit license but from what i saw, it is possible to work with that with the appropriate mph.options() command. 

Right now, it is impossible for me to change comsol version. Is there a way I could recieve some guidance on how to make it work? I understand you will need more information on how the installation/folders are placed.

**Comment by john-hen** [comment]
The first step is to run the "discovery" part of the test suite. It should look something like this:
```
❯ python tools\test.py discovery --log
Running test group "discovery".
[00:00.659] Searching system for available Comsol back-ends.
[00:00.659] Searching Windows Registry for Comsol executables.
[00:00.659] Checking registry node "SOFTWARE\Comsol\COMSOL55".
[00:00.659] Checking installation folder "C:\Program Files\Comsol55\Multiphysics".
[00:00.659] Found Comsol executable "C:\Program Files\Comsol55\Multiphysics\bin\win64\comsol.exe".
[00:00.675] Looking for Comsol executable on system search path.
[00:00.675] Running shell command "where comsol".
[00:00.769] Command exited with an error.
[00:00.769] Checking executable "C:\Program Files\Comsol55\Multiphysics\bin\win64\comsol.exe".
[00:00.769] Found Java bridge configuration "C:\Program Files\Comsol55\Multiphysics\bin\win64\comsol.ini".
[00:00.769] System architecture is "win64".
[00:00.769] Java VM at relative path "..\..\java\win64\jre\bin\server\jvm.dll".
[00:00.785] Root folder is "C:\Program Files\Comsol55\Multiphysics".
[00:01.570] Reported version information is "COMSOL Multiphysics 5.5.0.359".
[00:01.570] Assigned name "5.5" to this installation.
Passed in 2 s.
```

You'll have to download or `git clone` the source code from here on GitHub. I'd also recommend you uninstall your existing MPh and install from source in editable mode: `pip install -e .[dev]`. This means you can change the source code as you wish and use that modified version in your Python scripts. The `[dev]` extra installs tools such as pyTest, which is needed to run the test suite. You are looking for differences between your Comsol 5.2 and Comsol 5.5, used in the test above. Differences in terms of where files/folders are located inside the Comsol folder.

**Reply by Darkforg** [reply]
Thank you for your swift reply and help.

I will try to implement those changes and get back to you a.s.a.p.

If I understand correctly, the full command should be `pip install -e .[dev] mph`.

**Reply by Darkforg** [reply]
I tried running `python tools\test.py discovery --log` which in my files is slightly different (`MPh\tools\test.py`) which returns the following SyntaxError: invalid syntax
line 86  print(f'Running test group "{group}".')

Also, after git cloning the source code, `pip install -e .[dev]` is not working as a command for me. Is there something i am omitting?

**Reply by john-hen** [reply]
Is your Python version old too? Though it would have to be very old to balk at f-strings. But I don't see any other reason for that syntax error. What does `python --version` say? Same for `pip --version`. That command should work as it is.

**Reply by Darkforg** [reply]
I tried again in a newer python version (3.11.4) and `MPh\tools\test.py discovery --log` indeed runs. It does give a different kind of error:

line 6, in <module>
    from numpy import array, ndarray       # numerical array
ModuleNotFoundError: No module named 'numpy'

My pip version is 23.2.1, neither `pip install -e .[dev]` or `pip install -e .[dev] mph` works for me (i git cloned the code and uninstalled my mph afterwards)

**Reply by john-hen** [reply]
You need NumPy installed, but that's what the `pip install ...` command will do once it runs. It's `pip install -e .[dev]` (not: `... mph`). The `.` means "this directory here". I think you just need to `cd MPh`, based on how you called `test.py`. Pip will look for `pyproject.toml`, so that's the directory you need to be in.

**Reply by Darkforg** [reply]
Yes, you are correct. I am sorry, that was silly. I managed to run the discovery. should i dm the result or post here?

**Reply by john-hen** [reply]
Post here. Inside triple back-ticks ("code fences"). Like so:

````
```
[log output]
```
````

**Reply by Darkforg** [reply]
```
(base) C:\Users\Giannis\MPh>python tools\test.py discovery --log
Running test group "discovery".
[00:00.431] Searching system for available Comsol back-ends.
[00:00.431] Searching Windows Registry for Comsol executables.
[00:00.431] Checking registry node "SOFTWARE\Comsol\COMSOL52".
[00:00.432] Checking installation folder "C:\Program Files\COMSOL\COMSOL52\Multiphysics".
[00:00.432] Found Comsol executable "C:\Program Files\COMSOL\COMSOL52\Multiphysics\bin\win64\comsol.exe".
[00:00.432] Looking for Comsol executable on system search path.
[00:00.432] Running shell command "where comsol".
[00:00.535] Command exited with an error.
[00:00.535] Checking executable "C:\Program Files\COMSOL\COMSOL52\Multiphysics\bin\win64\comsol.exe".
[00:00.535] Found Java bridge configuration "C:\Program Files\COMSOL\COMSOL52\Multiphysics\bin\win64\comsol.ini".
[00:00.535] System architecture is "win64".
[00:00.540] Did not find Java VM path in "comsol.ini".
[00:00.540] Could not locate any Comsol installation.
Traceback (most recent call last):
  File "C:\Users\Giannis\MPh\tests\test_discovery.py", line 39, in <module>
    test_backend()
  File "C:\Users\Giannis\MPh\tests\test_discovery.py", line 25, in test_backend
    backend = mph.discovery.backend()
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Giannis\MPh\mph\discovery.py", line 415, in backend
    raise RuntimeError(error)
RuntimeError: Could not locate any Comsol installation.
Failed after 1 s.
```

**Reply by john-hen** [reply]
Okay, so it couldn't parse that `comsol.ini` file. I guess its format has changed somewhat since Comsol 5.2 (but not since Comsol 5.5). Here's what it looks like in 5.5:
```
# Configuration file for the Java bridge called from .NET
# '@user.dir' is evaluated to the same directory as this file
# '@user.home' is evaluated to the user's home directory
# '@os.home' is evaluated to the windows directory (e.g. 'c:\windows')
# '@comsol.root' is evaluated to the COMSOL root installation directory
-vm
..\..\java\win64\jre\bin\server\jvm.dll
-Djava.class.path=@comsol.root/plugins/com.comsol.webbridge_1.0.0.jar;@comsol.root/plugins/com.comsol.bridge_1.0.0.jar;@comsol.root/plugins/com.comsol.systemutils_1.0.0.jar;@comsol.root/plugins/org.eclipse.equinox.launcher_1.4.0.v20161219-1356.jar;@comsol.root/plugins/org.eclipse.osgi_3.12.100.v20180210-1608.jar

-Dcs.root=@comsol.root
-Dcs.logoutput=file
-Dcs.logfileprefix=javabridge
-Dcs.singleprocess=on
-Dinstall.dir=@comsol.root
-Djava.security.manager=
-Djava.security.policy=@comsol.root/bin/win64/comsol.policy
-Dcs.javasecurity=on
-Xss4m
-Xms40m
-Xmx2g
-XX:MaxMetaspaceSize=512m
```

MPh is looking for these two lines in particular:
```
-vm
..\..\java\win64\jre\bin\server\jvm.dll
```

More precisely, it looks for the line `-vm` and then grabs the relative path to the Java VM from the line that follows. That's the part that seems to fail. So then the question is, what does your `comsol.ini` look like?

**Reply by Darkforg** [reply]
```
# Configuration file for the Java bridge called from .NET
# '@user.dir' is evaluated to the same directory as this file
# '@user.home' is evaluated to the user's home directory
# '@os.home' is evaluated to the windows directory (e.g. 'c:\windows')
# '@comsol.root' is evaluated to the COMSOL root installation directory
-Djava.class.path=@comsol.root/plugins/com.comsol.webbridge_1.0.0.jar;@comsol.root/plugins/com.comsol.bridge_1.0.0.jar;@comsol.root/plugins/com.comsol.systemutils_1.0.0.jar;@comsol.root/plugins/org.eclipse.equinox.launcher_1.3.0.v20140415-2008.jar;@comsol.root/plugins/org.eclipse.osgi_3.10.1.v20140909-1633.jar

-Dcs.root=@comsol.root
-Dcs.logoutput=file
-Dcs.logfileprefix=javabridge
-Dcs.singleprocess=on
-Dinstall.dir=@comsol.root
-Djava.security.manager=
-Djava.security.policy=@comsol.root/bin/win64/comsol.policy
-Dcs.javasecurity=on
-Xss4m
-Xms40m
-Xmx1024m
-XX:MaxPermSize=256m
-Dcs.ckl
-Dcs.defaultserviceport=2036
```

**Reply by john-hen** [reply]
Well, there we go. These lines don't even exist. That explains that.

What you can try to do, is hard-code `jvm_path` in your "editable" installation of MPh. That is, edit `discovery.py`, find this piece of code:
```python
        # Parse Java bridge configuration.
        with ini.open(encoding='UTF-8') as stream:
            jvm_on_next_line = False
            for line in stream:
                line = line.strip()
                if not line or line.startswith('#'):
                    continue
                if jvm_on_next_line:
                    jvm_path = line
                    break
                if line.startswith('-vm'):
                    jvm_on_next_line = True
            else:
                log.debug(f'Did not find Java VM path in "{ini.name}".')
                continue
        log.debug(f'Java VM at relative path "{jvm_path}".')
```

Currently, that code block starts at line number 313. Replace the entire block with:
```python
        jvm_path = r'..\..\java\win64\jre\bin\server\jvm.dll'
```
at the same indentation level. Note the `r`-prefix before the string literal. It takes care of escaping the backslashes (`\`) . Though you could also replace the backslashes with forward slashes (`/`) and drop the `r`-prefix, Windows does not mind. It's Python that may interpret unescaped backslashes differently.

Obviously, I cannot test this. So I'm hoping the relative path is correct, even on Comsol 5.2. Maybe check in File Explorer if the `jvm.dll` actually exists at that location: relative to `comsol.ini`.

Unless I'm missing something, and I easily could, this should fix the issue of "locating the Comsol installation". I'm pretty sure the rest of the test suite won't pass (you only ran the "discovery" part), because the model we're testing against (`capacitor.mph` from the Tutorial) uses some features that weren't introduced until Comsol 5.4. But that doesn't mean the library as such won't work for you. If the discovery test now passes, then `mph.start()` should work as well. And you take it from there.

**Reply by Darkforg** [reply]
Yes `jvm.dll` is in the appropriate place and discovery runs:
```
(base) C:\Users\Giannis\MPh>python tools\test.py discovery --log
Running test group "discovery".
[00:00.230] Searching system for available Comsol back-ends.
[00:00.230] Searching Windows Registry for Comsol executables.
[00:00.231] Checking registry node "SOFTWARE\Comsol\COMSOL52".
[00:00.231] Checking installation folder "C:\Program Files\COMSOL\COMSOL52\Multiphysics".
[00:00.231] Found Comsol executable "C:\Program Files\COMSOL\COMSOL52\Multiphysics\bin\win64\comsol.exe".
[00:00.231] Looking for Comsol executable on system search path.
[00:00.231] Running shell command "where comsol".
[00:00.328] Command exited with an error.
[00:00.328] Checking executable "C:\Program Files\COMSOL\COMSOL52\Multiphysics\bin\win64\comsol.exe".
[00:00.328] Found Java bridge configuration "C:\Program Files\COMSOL\COMSOL52\Multiphysics\bin\win64\comsol.ini".
[00:00.328] System architecture is "win64".
[00:00.329] Root folder is "C:\Program Files\COMSOL\COMSOL52\Multiphysics".
[00:00.851] Reported version information is "COMSOL 5.2.0.220".
[00:00.851] Assigned name "5.2" to this installation.
Passed in 1 s.
```

The problem is when now when I try to run mph.start() i get this error:

`SystemError: java.lang.UnsupportedClassVersionError: org/jpype/classloader/DynamicClassLoader : Unsupported major.minor version 52.0`

Is it something I need to update?

**Reply by john-hen** [reply]
That doesn't look good. I don't know how to fix that. According to the error message, [it means](https://stackoverflow.com/q/22489398) there's a mismatch between the Java versions used at compile time vs. at run time. But that doesn't make sense since we're using the Java run-time environment that ships with Comsol, in order to avoid this very problem. So then maybe it's a problem with JPype (the Python-to-Java bridge we're using), but we have no control over that. It either works or it doesn't. Perhaps Comsol 5.2 shipped with a Java version that JPype does not or does no longer support.

For the record, this is the Java version Comsol 5.5 uses:
```
> cd "C:\Program Files\COMSOL\COMSOL55\Multiphysics\java\win64\jre\bin"
> java -version
openjdk version "1.8.0_222"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_222-b10)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.222-b10, mixed mode)
```

So that's Java 8.

**Reply by Darkforg** [reply]
Unfortunately it runs on java 7
```
(base) C:\Program Files\COMSOL\COMSOL52\Multiphysics\java\win64\jre\bin>java -version
java version "1.7.0_75"
Java(TM) SE Runtime Environment (build 1.7.0_75-b13)
Java HotSpot(TM) 64-Bit Server VM (build 24.75-b04, mixed mode)
```
Is there nothing we can do about that?

**Reply by john-hen** [reply]
Turns out, Java 7 is definitely the problem. JPype is built with Java 8, which explains the above error (jpype-project/jpype#1083). Older JPype versions may have supported earlier Java versions, but not since at least JPype 0.7.5 (jpype-project/jpype#742). And I wouldn't downgrade JPype that far, it just invites a whole bunch of other issues and will most likely not work.

The other options are:
* Build the current JPype 1.4.1 from source using a Java 7 development kit, then pip-install that.
* Run Comsol with an external Java 8 environment and hard-code its location in MPh.

They both have a decent chance of working. No guarantee obviously. I don't know how much further you want to go with this, it'll cost some time.

**Reply by Darkforg** [reply]
> The other options are:
> Build the current JPype 1.4.1 from source using a Java 7 development kit, then pip-install that.

I am willing to give that a go if you could give me the general direction. I do understand I have already taken too much of your time and I would not want to impose.

**Reply by john-hen** [reply]
Instructions for building JPype from source can be found in its documentation: https://jpype.readthedocs.io/en/latest/install.html#build-and-install-manually

But first you'll need to install Java and C++ compilers. The latter means, get Visual Studio (the free Community Edition will do) and install the C++ tool chain. And you need a Java 7 development kit (JDK 7). I would try maybe `winget install -i Azul.Zulu.7.JDK` (but I haven't tried), otherwise google to find an installer. Obviously Java 7 is pretty old by now, so won't be that easily available anymore. Make sure you're setting the `JAVA_HOME` environment variable. Then compile JPype as per the docs.

**Reply by Darkforg** [reply]
Thank you for your invaluable help.

I will try to do it and contact you for the results a.s.a.p.

**Reply by john-hen** [reply]
I looked into this a little bit today. Unfortunately, I can't get any (reasonably recent) version of JPype to build with Java 7. So I guess when Thrameos (the JPype developer) said "It may still be possible to compile a library for the older Java versions", he was being optimistic. There is a Java 8 syntax feature (the double-colon operator, `::`) that JPype has embraced quite a bit. And I, for one, do not know how to back-port all that to equivalent Java 7 syntax. So I think this is a dead end.

If you want to pursue the other option (running Comsol with an external Java 8 run-time environment), you'd have to patch the `backend()` function in `discovery.py` to return a dictionary with the correct values filled in. Something like:
```python
def backend(version=None):
    build = 0   #  I don't know actual the build number, but it shouldn't matter.
    root = Path(r'C:\Program Files\COMSOL\COMSOL52\Multiphysics')
    jvm = Path('...')  # absolute path to the external Java 8 `jvm.dll`
    server = root/'bin'/'win64'/'comsolmphserver.exe'
    return  {
      'name':   '5.2',
      'major':  5,
      'minor':  2,
      'patch':  0,
      'build':  build,  
      'root':   root,
      'jvm':    jvm,
      'server': server,
    }
```
Everything else in `discovery.py` is then irrelevant and could even be removed. I'd install AdoptOpenJDK 8, as that's what Comsol uses in later versions, and it should be backward-compatible.

**Reply by Darkforg** [reply]
Thank you for the new information. Unfortunately I haven't managed to look at it yet due to other commitments. I will certainly look into it and notify you when there is something to share. It would be great if I managed to make a useful contribution.
